lxd/images.go

```diff
@@ -1452,30 +1452,46 @@ func getImageMetadata(fname string) (*api.ImageMetadata, string, error) {
 	return &result, imageType, nil
 }
 
-func doImagesGet(ctx context.Context, tx *db.ClusterTx, recursion bool, projectName string, public bool, clauses *filter.ClauseSet, hasPermission auth.PermissionChecker) (any, error) {
+func doImagesGet(ctx context.Context, tx *db.ClusterTx, recursion bool, projectName string, public bool, clauses *filter.ClauseSet, hasPermission auth.PermissionChecker, allProjects bool) (any, error) {
 	mustLoadObjects := recursion || (clauses != nil && len(clauses.Clauses) > 0)
 
-	fingerprints, err := tx.GetImagesFingerprints(ctx, projectName, public)
-	if err != nil {
-		return err, err
+	imagesProjectsMap := map[string][]string{}
+	if allProjects {
+		var err error
+
+		imagesProjectsMap, err = tx.GetImages(ctx)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		fingerprints, err := tx.GetImagesFingerprints(ctx, projectName, public)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, fp := range fingerprints {
+			imagesProjectsMap[fp] = []string{projectName}
+		}
 	}
 
 	var resultString []string
 	var resultMap []*api.Image
 
 	if recursion {
-		resultMap = make([]*api.Image, 0, len(fingerprints))
+		resultMap = make([]*api.Image, 0, len(imagesProjectsMap))
 	} else {
-		resultString = make([]string, 0, len(fingerprints))
+		resultString = make([]string, 0, len(imagesProjectsMap))
 	}
 
-	for _, fingerprint := range fingerprints {
-		image, err := doImageGet(ctx, tx, projectName, fingerprint, public)
+	for fingerprint, projects := range imagesProjectsMap {
+		curProjectName := projects[0]
+
+		image, err := doImageGet(ctx, tx, curProjectName, fingerprint, public)
 		if err != nil {
 			continue
 		}
 
-		if !image.Public && !hasPermission(entity.ImageURL(projectName, fingerprint)) {
+		if !image.Public && !hasPermission(entity.ImageURL(curProjectName, fingerprint)) {
```

Comment:
Also, `GetImages()` only returns the image fingerprints. When we iterate over the fingerprints and projects, `doImageGet()` is called with the `public` argument passed in. Maybe I'm missing something here?

Reply:
I think I'm following what you mean now. Maybe we can rework `GetImages()` to include a `public` parameter, similar to `GetImagesFingerprints()`, like so:

https://github.com/canonical/lxd/blob/9ac2433510c825833f260621b00fa0c19e6a6ff8/lxd/db/images.go#L183-L192

Reply:
I think we should remove `GetImagesFingerprints` entirely. We're using it to get a list of fingerprints and then re-querying the same table to get values from the remaining columns. Again though I think this can come in a later refactor. For now I'd be happy if getting images from all projects follows the same pattern and only returns public images when the caller is not authenticated.

Reply:
Just had a thought. Now that we're disallowing `all-projects` requests for untrusted users we don't need to do any of the above refactoring here, since we'll always be authenticated when getting images across all projects :)